gems/sorbet-runtime/test/types/final_method.rb

```diff
@@ -0,0 +1,319 @@
+        def self.foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be redefined")
+  end
+
+  it "forbids redefining a final instance method with no sig" do
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        extend T::Sig
+        sig(:final) {void}
+        def foo; end
+        def foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be redefined")
+  end
+
+  it "forbids redefining a final class method with no sig" do
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        extend T::Sig
+        sig(:final) {void}
+        def self.foo; end
+        def self.foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be redefined")
+  end
+
+  it "allows redefining a regular instance method to be final" do
+    Class.new do
+      extend T::Sig
+      def foo; end
+      sig(:final) {void}
+      def foo; end
+    end
+  end
+
+  it "allows redefining a regular class method to be final" do
+    Class.new do
+      extend T::Sig
+      def self.foo; end
+      sig(:final) {void}
+      def self.foo; end
+    end
+  end
+
+  it "forbids overriding a final instance method" do
+    c = Class.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new(c) do
+        def foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final class method" do
+    c = Class.new do
+      extend T::Sig
+      sig(:final) {void}
+      def self.foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new(c) do
+        def self.foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method from an included module" do
+    m = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        include m
+        def foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method from an extended module" do
+    m = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        extend m
+        def self.foo; end
+      end
+    end
+    assert_includes(err.message, "was declared as final and cannot be overridden")
+  end
+
+  it "forbids overriding a final method by including two modules" do
+    m1 = Module.new do
+      extend T::Sig
+      sig(:final) {void}
+      def foo; end
+    end
+    m2 = Module.new do
+      def foo; end
+    end
+    err = assert_raises(RuntimeError) do
+      Class.new do
+        include m2, m1
```

Comment:
Why is this `m2, m1` and not the other way around?

Reply:
Oh wow, that's interesting. Calling `super` in `m2#foo` will call `m1#foo`, which if it's final is essentially an override. That's kind of nuts.

Reply:
The include order is backwards. So `include A, B` is like `include B; include A`.